enterprise/server/oci/ocifetcher/ocifetcher.go

```diff
@@ -0,0 +1,181 @@
+}
+
+func NewClient(allowedPrivateIPs []*net.IPNet, mirrors []interfaces.MirrorConfig) ofpb.OCIFetcherClient {
+	return &ociFetcherClient{
+		allowedPrivateIPs: allowedPrivateIPs,
+		mirrors:           mirrors,
+	}
+}
+
+// NewMirrorTransport wraps an http.RoundTripper with registry mirror support.
+// Requests matching a mirror's OriginalURL are rewritten to the MirrorURL,
+// with automatic fallback to the original URL on failure.
+func NewMirrorTransport(inner http.RoundTripper, mirrors []interfaces.MirrorConfig) http.RoundTripper {
+	return &mirrorTransport{
+		inner:   inner,
+		mirrors: mirrors,
+	}
+}
+
+// verify that mirrorTransport implements the RoundTripper interface.
+var _ http.RoundTripper = (*mirrorTransport)(nil)
+
+type mirrorTransport struct {
+	inner   http.RoundTripper
+	mirrors []interfaces.MirrorConfig
+}
+
+func (t *mirrorTransport) RoundTrip(in *http.Request) (out *http.Response, err error) {
+	for _, mirror := range t.mirrors {
+		if match, err := matchesMirror(mirror, in.URL); err == nil && match {
+			mirroredRequest, err := rewriteToMirror(mirror, in)
+			if err != nil {
+				log.Errorf("error mirroring request: %s", err)
+				continue
+			}
+			out, err := t.inner.RoundTrip(mirroredRequest)
+			if err != nil {
+				log.Errorf("mirror err: %s", err)
+				continue
+			}
+			if out.StatusCode < http.StatusOK || out.StatusCode >= 300 {
+				fallbackRequest, err := rewriteFallback(mirror, in)
+				if err != nil {
+					log.Errorf("error rewriting fallback request: %s", err)
+					continue
+				}
+				return t.inner.RoundTrip(fallbackRequest)
+			}
+			return out, nil
+		}
+	}
+	return t.inner.RoundTrip(in)
+}
+
+func matchesMirror(mc interfaces.MirrorConfig, u *url.URL) (bool, error) {
+	originalURL, err := url.Parse(mc.OriginalURL)
+	if err != nil {
+		return false, err
+	}
+	return originalURL.Host == u.Host, nil
+}
+
+func rewriteToMirror(mc interfaces.MirrorConfig, originalRequest *http.Request) (*http.Request, error) {
+	mirrorURL, err := url.Parse(mc.MirrorURL)
+	if err != nil {
+		return nil, err
+	}
+	originalURL := originalRequest.URL.String()
+	req := originalRequest.Clone(originalRequest.Context())
+	req.URL.Scheme = mirrorURL.Scheme
+	req.URL.Host = mirrorURL.Host
+	// Set X-Forwarded-Host so the mirror knows which remote registry to make requests to.
+	// ociregistry looks for this header and will default to forwarding requests to Docker Hub if not found.
+	req.Header.Set("X-Forwarded-Host", originalRequest.URL.Host)
+	log.Debugf("%q rewritten to %s", originalURL, req.URL.String())
+	return req, nil
+}
+
+func rewriteFallback(mc interfaces.MirrorConfig, originalRequest *http.Request) (*http.Request, error) {
+	originalURL, err := url.Parse(mc.OriginalURL)
+	if err != nil {
+		return nil, err
+	}
+	req := originalRequest.Clone(originalRequest.Context())
+	req.URL.Scheme = originalURL.Scheme
+	req.URL.Host = originalURL.Host
+	log.Debugf("(fallback) %q rewritten to %s", originalURL, req.URL.String())
```

Comment:
nit: would this log be more useful with `log.Ctx*`? (I guess maybe `originalRequest.Context()` might work / be convenient?)